# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fetch/InternalRequest.cpp
# Commit: ea27cd66fefd
# Full Hash: ea27cd66fefd175c69de110191319544a252e820
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2023-01-18 09:51:31
# Regressor Bug: 1351231
# File Overlap Count: 1
# Description:
#   Bug 1351231 - Support conversation between InternalRequest and IPCInternalRequest. r=dom-worker-reviewers,jesup
#   
#   Depends on D138812
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138813
# ==============================================================================

diff -r 8c21c9468949 -r ea27cd66fefd dom/fetch/InternalRequest.cpp
--- a/dom/fetch/InternalRequest.cpp	Tue Jan 17 15:13:44 2023 +0000
+++ b/dom/fetch/InternalRequest.cpp	Tue Jan 17 15:13:44 2023 +0000
@@ -53,6 +53,7 @@
 
   copy->mPreferredAlternativeDataType = mPreferredAlternativeDataType;
   copy->mSkipWasmCaching = mSkipWasmCaching;
+  copy->mEmbedderPolicy = mEmbedderPolicy;
   return copy;
 }
 
@@ -140,6 +141,7 @@
       mUnsafeRequest(aOther.mUnsafeRequest),
       mUseURLCredentials(aOther.mUseURLCredentials),
       mContentPolicyTypeOverridden(aOther.mContentPolicyTypeOverridden),
+      mEmbedderPolicy(aOther.mEmbedderPolicy),
       mInterceptionContentPolicyType(aOther.mInterceptionContentPolicyType),
       mInterceptionRedirectChain(aOther.mInterceptionRedirectChain),
       mInterceptionFromThirdParty(aOther.mInterceptionFromThirdParty) {
@@ -163,12 +165,14 @@
           static_cast<nsContentPolicyType>(aIPCRequest.contentPolicyType())),
       mReferrer(aIPCRequest.referrer()),
       mReferrerPolicy(aIPCRequest.referrerPolicy()),
+      mEnvironmentReferrerPolicy(aIPCRequest.environmentReferrerPolicy()),
       mMode(aIPCRequest.requestMode()),
       mCredentialsMode(aIPCRequest.requestCredentials()),
       mCacheMode(aIPCRequest.cacheMode()),
       mRedirectMode(aIPCRequest.requestRedirect()),
       mIntegrity(aIPCRequest.integrity()),
       mFragment(aIPCRequest.fragment()),
+      mEmbedderPolicy(aIPCRequest.embedderPolicy()),
       mInterceptionContentPolicyType(static_cast<nsContentPolicyType>(
           aIPCRequest.interceptionContentPolicyType())),
       mInterceptionRedirectChain(aIPCRequest.interceptionRedirectChain()),
@@ -186,11 +190,52 @@
 
   const Maybe<BodyStreamVariant>& body = aIPCRequest.body();
 
-  // This constructor is (currently) only used for parent -> child communication
-  // (constructed on the child side).
-  if (body) {
-    MOZ_ASSERT(body->type() == BodyStreamVariant::TParentToChildStream);
-    mBodyStream = body->get_ParentToChildStream().stream();
+  if (!XRE_IsParentProcess()) {
+    if (body) {
+      MOZ_ASSERT(body->type() == BodyStreamVariant::TParentToChildStream);
+      mBodyStream = body->get_ParentToChildStream().stream();
+    }
+  } else {
+    if (body) {
+      MOZ_ASSERT(body->type() == BodyStreamVariant::TChildToParentStream);
+      mBodyStream =
+          DeserializeIPCStream(body->get_ChildToParentStream().stream());
+    }
+  }
+}
+
+void InternalRequest::ToIPCInternalRequest(
+    IPCInternalRequest* aIPCRequest, mozilla::ipc::PBackgroundChild* aManager) {
+  aIPCRequest->method() = mMethod;
+  for (const auto& url : mURLList) {
+    aIPCRequest->urlList().AppendElement(url);
+  }
+  mHeaders->ToIPC(aIPCRequest->headers(), aIPCRequest->headersGuard());
+  aIPCRequest->bodySize() = mBodyLength;
+  aIPCRequest->preferredAlternativeDataType() = mPreferredAlternativeDataType;
+  aIPCRequest->contentPolicyType() = mContentPolicyType;
+  aIPCRequest->referrer() = mReferrer;
+  aIPCRequest->referrerPolicy() = mReferrerPolicy;
+  aIPCRequest->environmentReferrerPolicy() = mEnvironmentReferrerPolicy;
+  aIPCRequest->requestMode() = mMode;
+  aIPCRequest->requestCredentials() = mCredentialsMode;
+  aIPCRequest->cacheMode() = mCacheMode;
+  aIPCRequest->requestRedirect() = mRedirectMode;
+  aIPCRequest->integrity() = mIntegrity;
+  aIPCRequest->fragment() = mFragment;
+  aIPCRequest->embedderPolicy() = mEmbedderPolicy;
+
+  if (mPrincipalInfo) {
+    aIPCRequest->principalInfo() = Some(*mPrincipalInfo);
+  }
+
+  if (mBodyStream) {
+    nsCOMPtr<nsIInputStream> body = mBodyStream;
+    aIPCRequest->body().emplace(ChildToParentStream());
+    DebugOnly<bool> ok = mozilla::ipc::SerializeIPCStream(
+        body.forget(), aIPCRequest->body()->get_ChildToParentStream().stream(),
+        /* aAllowLazy */ false);
+    MOZ_ASSERT(ok);
   }
 }
 